File: app/src/main/java/UTSSG/ZQDesigned/Characterscode/MainActivity.java
package UTSSG.ZQDesigned.Characterscode;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import java.io.UnsupportedEncodingException;

import UTSSG.ZQDesigned.Characterscode.Utils.FunctionsToolset;

public class MainActivity extends AppCompatActivity {

    private static final String appVersionname = BuildConfig.VERSION_NAME;
    private static final int appVersioncode = BuildConfig.VERSION_CODE;


    private long exitTime = 0;

    @SuppressLint("SetTextI18n")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        TextView textView = findViewById(R.id.main_text_version);
        textView.setText("版本号：" + appVersionname + "(" + appVersioncode + ")");
        if (FunctionsToolset.isInstalled("com.android.google.gms")) {
            new AlertDialog.Builder(this).setTitle("提示")//设置对话框标题
                    .setMessage("GMS Detected！")//设置显示的内容
                    .setNegativeButton("返回", null)
                    .setIcon(android.R.drawable.ic_dialog_alert)//设置图标
                    .setCancelable(false)//设置点击对话框外部不消失
                    .show();//在按键响应事件中显示此对话框
        }
        }


        @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            exit();
            return false;
        }
        return super.onKeyDown(keyCode, event);
    }

    public void exit() {
        if ((System.currentTimeMillis() - exitTime) > 2500) {
            Toast.makeText(getApplicationContext(),"再按一次返回退出程序",Toast.LENGTH_SHORT).show();
            exitTime = System.currentTimeMillis();
        } else {
            finish();
        }
    }

    public void entrance_onClick(View v) {
        startActivity(new Intent(getApplicationContext(),MigrateActivity.class));
    }

    @SuppressLint("SetTextI18n")
    public void button2_Onclick(View view) throws UnsupportedEncodingException {
        EditText name = findViewById(R.id.name);
        String nameStr = name.getText().toString();
        
        if(TextUtils.isEmpty(nameStr)) {//判断
            Toast.makeText(this, "请输入内容！", Toast.LENGTH_SHORT).show();
            new AlertDialog.Builder(this).setTitle("提示")//设置对话框标题
                    .setMessage("请输入内容！")//设置显示的内容
                    .setNegativeButton("返回", null)
                    .setIcon(android.R.drawable.ic_dialog_alert)//设置图标
                    .setCancelable(false)//设置点击对话框外部不消失
                    .show();//在按键响应事件中显示此对话框
            return;
        }
        
        TextView result = findViewById(R.id.output);
        byte[] codeBit = nameStr.getBytes("gb2312");// 获取汉字的字节数组
        if(codeBit.length%2==0) {//判断
                    StringBuilder sb = new StringBuilder();
                    for (byte codeBit_tmp:codeBit) {
                        codeBit_tmp -= 160;
                        
                        if (codeBit_tmp < 0) {
                            Toast.makeText(this, "您输入的内容不合法！", Toast.LENGTH_SHORT).show();
                            new AlertDialog.Builder(this).setTitle("提示")//设置对话框标题
                                    .setMessage("您输入的内容不合法！")//设置显示的内容
                                    .setNegativeButton("返回", (dialog, which) -> {
                                    })
                                    .setIcon(android.R.drawable.ic_dialog_alert)//设置对话框图标
                                    .setCancelable(false)//设置点击对话框外部不消失
                                    .show();//在按键响应事件中显示此对话框
                            return;
                        } else if (codeBit_tmp <= 10) {
                            sb.append(0);
                        }
                        sb.append(codeBit_tmp);
                    }
                    Toast.makeText(this, "正在查询......", Toast.LENGTH_SHORT).show();
                    result.setText(nameStr+"\n↓↓转换结果↓↓\n"+(sb.toString().replaceAll(".{4}", "$0 ")));
        }else{
            Toast.makeText(this, "您输入的内容不合法！", Toast.LENGTH_SHORT).show();
            new AlertDialog.Builder(this).setTitle("提示")//设置对话框标题
                    .setMessage("您输入的内容不合法！")//设置显示的内容
                    .setNegativeButton("返回", null)
                    .setIcon(android.R.drawable.ic_dialog_alert)//设置对话框图标
                    .setCancelable(false)//设置点击对话框外部不消失
                    .show();//在按键响应事件中显示此对话框
        }

        }
        public void button1_Onclick(View view) {
            startActivity(new Intent(getApplicationContext(),DonateActivity.class));
        }

    }

File: MorseCode/src/main/java/UTSSG/ZQDesigned/MorseCode/MainActivity.kt
package UTSSG.ZQDesigned.MorseCode

import android.app.Activity
import android.content.SharedPreferences
import android.os.Bundle
import android.os.Looper
import android.view.View
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast

class MainActivity : Activity() {
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Toast.makeText(this, "Welcome!", Toast.LENGTH_SHORT).show()
        val sp = getSharedPreferences("MorseCode", MODE_PRIVATE)
        if (sp.getString("A", null) == null) {
            Toast.makeText(this, "首次运行，正在初始化存储......", Toast.LENGTH_SHORT).show()
            sp.lazySp {
                //向SharedPreferences写入所有字母和数字对应的摩斯电码
                putString("A", ".-")
                putString("B", "-...")
                putString("C", "-.-.")
                putString("D", "-..")
                putString("E", ".")
                putString("F", "..-.")
                putString("G", "--.")
                putString("H", "....")
                putString("I", "..")
                putString("J", ".---")
                putString("K", "-.-")
                putString("L", ".-..")
                putString("M", "--")
                putString("N", "-.")
                putString("O", "---")
                putString("P", ".--.")
                putString("Q", "--.-")
                putString("R", ".-.")
                putString("S", "...")
                putString("T", "-")
                putString("U", "..-")
                putString("V", "...-")
                putString("W", ".--")
                putString("X", "-..-")
                putString("Y", "-.--")
                putString("Z", "--..")
                putString("1", ".----")
                putString("2", "..---")
                putString("3", "...--")
                putString("4", "....-")
                putString("5", ".....")
                putString("6", "-....")
                putString("7", "--...")
                putString("8", "---..")
                putString("9", "----.")
                putString("0", "-----")
                putString(" ", " ")
            }
            Toast.makeText(this, "初始化完成！", Toast.LENGTH_SHORT).show()
        }
    }

    fun onClick(v: View) {
        val originaltext = findViewById<EditText>(R.id.originaltext)
        val input = originaltext.text.toString()
        //判断输入是否为空
        if (input.isEmpty()) {
            Toast.makeText(this, "请输入内容！", Toast.LENGTH_SHORT).show()
            return
        }
        //判断输入是否为纯英文或纯中文（忽略数字），如果不是则返回
        if (!input.isEnglish()) {
            Toast.makeText(this, "请输入纯英文或纯中文！", Toast.LENGTH_SHORT).show()
            return
        }
        //将输入的内容逐字拆分并写入数组
        val inputArray = input.toCharArray()
        //将数组内全部小写字母转换为大写字母
        for (i in inputArray.indices) {
            if (inputArray[i] in 'a'..'z') {
                inputArray[i] = inputArray[i].uppercaseChar()
            }
        }
        //从SharedPreferences中读取摩斯电码
        val sp = getSharedPreferences("MorseCode", MODE_PRIVATE)
        //将摩斯电码写入数组
        val morseArray = Array(inputArray.size) { sp.getString(inputArray[it].toString(), null) }
        //将摩斯电码数组转换为字符串
        val morseString = morseArray.joinToString(" ")
        //将摩斯电码字符串写入Result
        val result: TextView = findViewById(R.id.result)
        result.text = morseString
    }

    private inline fun SharedPreferences.lazySp(crossinline block: SharedPreferences.Editor.() -> Unit) {
        Thread {
            Looper.prepare()
            this.edit().apply {
                block()
            }.apply()
        }.start()
    }

    /**
     * 判断给定内容是否为纯英文(忽略数字)
     *
     * @return true为纯英文，false为非纯英文
     */
    private fun String.isEnglish(): Boolean {
        val inputArray = this.toCharArray()
        for (i in inputArray.indices) {
            if (inputArray[i] !in 'a'..'z' && inputArray[i] !in 'A'..'Z' && inputArray[i] !in '0'..'9' && inputArray[i] != ' ') {
                return false
            }
        }
        return true
    }

}